cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -748,11 +749,15 @@ class TableSorter {
     auto& comparator = comparator_;
     const auto& first_sort_key = sort_keys_[0];
 
+    ChunkLocation left_loc{0, 0};
+    ChunkLocation right_loc{0, 0};
     std::merge(nulls_begin, nulls_middle, nulls_middle, nulls_end, temp_indices,
                [&](uint64_t left, uint64_t right) {
                  // First column is either null or nan
-                 const auto left_loc = left_resolver_.Resolve(left);
-                 const auto right_loc = right_resolver_.Resolve(right);
+                 left_loc =
+                     left_resolver_.ResolveWithChunkIndexHint(left, left_loc.chunk_index);
```

Comment:
The point of this is to avoid all atomic accesses, right?

Reply:
Memory access really: keeping the `ChunkLocation::chunk_index` in registers throughout the loop -- of course it depends on the register allocation and inlining of `ResolveWithChunkIndexHint`. It improved the sort benchmarks slightly.